src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/LoginHandler.cs

```diff
@@ -0,0 +1,636 @@
+
+            byte[] rentedSSPIBuff = null;
+            byte[] outSSPIBuff = null; // track the rented buffer as a separate variable in case it is updated via the ref parameter
+            uint outSSPILength = 0;
+
+            // only add lengths of password and username if not using SSPI or requesting federated authentication info
+            if (!context.UseSspi && !(context.Features.FederatedAuthenticationInfoRequested || context.Features.FederatedAuthenticationRequested))
+            {
+                checked
+                {
+                    length += (userName.Length * 2) + encryptedPasswordLengthInBytes
+                    + encryptedChangePasswordLengthInBytes;
+                }
+            }
+            else
+            {
+                if (context.UseSspi)
+                {
+                    throw new NotImplementedException("SSPI is not implemented");
+                }
+            }
+
+
+            int feOffset = length;
+            // calculate and reserve the required bytes for the featureEx
+            length = CalculateFeatureExtensionLength(context);
+
+            length += feOffset;
+
+            // TODO : Plumb Session recovery.
+            SessionData recoverySessionData = null;
+
+            await WriteLoginData(context,
+                           recoverySessionData,
+                           encryptedPassword,
+                           encryptedChangePassword,
+                           encryptedPasswordLengthInBytes,
+                           encryptedChangePasswordLengthInBytes,
+                           userName,
+                           length,
+                           feOffset,
+                           context.ClientInterfaceName,
+                           outSSPIBuff,
+                           outSSPILength,
+                           isAsync,
+                           ct).ConfigureAwait(false);
+
+            if (rentedSSPIBuff != null)
+            {
+                ArrayPool<byte>.Shared.Return(rentedSSPIBuff, clearArray: true);
+            }
+
+            TdsStream stream = context.TdsStream;
+            ct.ThrowIfCancellationRequested();
+            
+            if (isAsync)
+            {
+                await stream.FlushAsync(ct).ConfigureAwait(false);
+            }
+            else
+            {
+                stream.Flush();
+            }
+        }
+
+
+        /// <summary>
+        /// Calculates the length in bytes to send all the feature extensions.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        private int CalculateFeatureExtensionLength(LoginHandlerContext context)
+        {
+            int feLength = 0;
+            TdsEnums.FeatureExtension requestedFeatures = context.Features.RequestedFeatures;
+            if (context.UseFeatureExt)
+            {
+                if (_sessionRecoveryFeature.ShouldUseFeature(requestedFeatures))
```

Comment:
Same feedback for this block, ShouldUseFeature could be internally handled by GetLengthInBytes,

Reply:
I want to keep them separate.
I agree that GetLengthInBytes can check for "ShouldUseFeature"
But then that overloads the return value of Getting the length. 
A feature can have 0 byte length by design if there is no data exchange in the feature extension. 

I want to make the intention clear with a step which checks if feature should be used and it is a 0 byte feature vs if it is 0 byte because it was not requested.